openedx/core/lib/api/parsers.py

```diff
@@ -1,7 +1,90 @@
+for details.
+"""
 
-class MergePatchParser(parsers.JSONParser):
+from django.utils.translation import ugettext as _
+
+from rest_framework.exceptions import ParseError, UnsupportedMediaType
+from rest_framework.parsers import FileUploadParser, JSONParser
+
+
+class TypedFileUploadParser(FileUploadParser):
+    """
+    Handles upload of files, ensuring that the media type is supported, and
+    that the uploaded filename matches the Content-type.
+
+    Requirements:
+        * The view must have an `upload_media_types` attribute which is a
+          set (or other container) enumerating the mimetypes of the supported
+          media formats
+
+          Example:
+
+              View.upload_media_types = {'audio/mp3', 'audio/ogg', 'audio/wav'}
+
+        * Content-type must be set to a supported type (as
+          defined in View.upload_media_types above).
+
+          Example:
+
+              Content-type: audio/ogg
+
+        * Content-disposition must include a filename with a valid extension
+          for the specified Content-type.
+
+          Example:
+
+              Content-disposition: attachment; filename="lecture-1.ogg"
+    """
+
+    media_type = '*/*'
+
+    # Add more entries to this as needed.  All extensions should be lowercase.
+    file_extensions = {
+        'image/gif': {'.gif'},
+        'image/jpeg': {'.jpeg', '.jpg'},
+        'image/pjpeg': {'.jpeg', '.jpg'},
+        'image/png': {'.png'},
+        'image/svg': {'.svg'},
+    }
+
+    def parse(self, stream, media_type=None, parser_context=None):
+        """
+        Parse the request, returning a DataAndFiles object with the data dict
+        left empty, and the body of the request placed in files['file'].
+        """
+
+        upload_media_types = getattr(parser_context['view'], 'upload_media_types', None)
+        if upload_media_types is not None and media_type not in upload_media_types:
+            raise UnsupportedMediaType(media_type)
+
+        filename = self.get_filename(stream, media_type, parser_context)
+        if media_type in self.file_extensions:
+            fileparts = filename.rsplit('.', 1)
+            if len(fileparts) < 2:
+                ext = ''
+            else:
+                ext = '.{}'.format(fileparts[1])
+            if ext.lower() not in self.file_extensions[media_type]:
+                errmsg = (
+                    u"File extension does not match requested Content-type. "
+                    u"Filename: {filename}, Content-type: {contenttype}"
+                )
+                usermsg = _(u"Unable to parse file {filename}.  A problem occurred with the file extension or type.")
+                raise ParseError(detail={
+                    'error_code': 'invalid_file_extension',
+                    'developer_message': errmsg.format(filename=filename, contenttype=media_type),
+                    'user_message': usermsg.format(filename=filename),
+                })
```

Comment:
This doesn't do what I want.  The returned response body is 

    {"detail":"{'developer_message': u'File extension does not match requested Content-type. Filename: file.png, Content-type: image/jpeg', 'error_code': 'invalid_file_extension', 'user_message': u'Unable to parse file file.png.  A problem occurred with the file extension or type.'}"}

That is, it's a json response with one key, "detail", whose value is the repr of the python dictionary I created. I'm looking to get the python dictionary as the response body:

    {"developer_message": "File extension does not match requested Content-type. Filename: file.png, Content-type: image/jpeg", "error_code": "invalid_file_extension", "user_message": "Unable to parse file file.png.  A problem occurred with the file extension or type."}

Is there a good way to get responses raised as exceptions to return the right format?  Because the thing getting returned from the parser isn't a response, so we can't just build the response from scratch.  I notice DRF has an `EXCEPTION_HANDLER` setting that would let us override the response returned by APIExceptions like this.  

I could easily write a handler that overrides the response if detail is set to a dict, but I don't know if that would cause troubles elsewhere.